### DISCUSSION TITLE
How to integrate libuv with libraries like curl and MySQL?

### DISCUSSION THREAD

**Discussion by HallofFamer** [libuv/libuv#4427] [General]
I have been able to get libuv to work for timer, file system and most network operations, though these have been libuv’s builtin functionality without interacting with third party libraries. Is there a tutorial on how libuv works works with curl and similar libraries? 

Lets say I want to make two http get requests using curl, before I would just run them in series with two curl_easy_perform. With libuv, is it possible to run them in non blocking manner and execute callbacks when each request is complete?

**Comment by cjihrig** [comment]
I'm not sure if there is a tutorial, but you could look at [txiki.js](https://github.com/saghul/txiki.js), which does integrate curl and libuv (and is a project of a libuv maintainer).

> With libuv, is it possible to run them in non blocking manner and execute callbacks when each request is complete?

You could look at the [threadpool documentation](https://docs.libuv.org/en/v1.x/threadpool.html). The libuv test suite should also have a number of examples covering this.

**Comment by santigimeno** [comment]
Here's a good example from the curl documentation: https://curl.se/libcurl/c/multi-uv.html

**Reply by saghul** [reply]
That's pretty much how I did it for txiki.js indeed!

**Reply by HallofFamer** [reply]
Hi @santigimeno 

Thx for linking me to this, I have a question regarding this. Is it absolutely necessary to specify a timer function(with uv_timer_t) for libuv to work with curl? I tried to only set option for socket function, and it would not be invoked at all. Seems the timer has to be set, or is there a workaround to this? 

Then I tried to set up timer function, and instead it kept giving me this error: `Exception thrown: read access violation. loop was 0xFFFFFFFFFFFFFFA7`. It points to the source code of `timer_heap` function. 

```C
static struct heap *timer_heap(const uv_loop_t* loop) {
#ifdef _WIN32
  return (struct heap*) loop->timer_heap;
#else
  return (struct heap*) &loop->timer_heap;
#endif
}
```

Strangely enough my event loop has been initialized correctly, i have no idea why it causes a segfault with read access violation. The only difference I had from the example code is that I assigned `uv_timer_t` as timer data and it was passed to the start_timeout function as userdata void pointer, since my code does not use global variables.

**Reply by saghul** [reply]
Can you paste the code of how you're calling it? Did you allocate the timer on the heap?

**Reply by HallofFamer** [reply]
@saghul 

This is the code that I wrote to setopt for curl multi handle: 

```
    uv_timer_t* timerHandle = (uv_timer_t*)malloc(sizeof(uv_timer_t));
    uv_timer_init(eventLoop, timerHandle);
    curl_multi_setopt(curlMultiHandle, CURLMOPT_SOCKETFUNCTION, httpPollSocket);
    curl_multi_setopt(curlMultiHandle, CURLMOPT_TIMERFUNCTION, httpStartTimeout);
    curl_multi_setopt(curlMultiHandle, CURLMOPT_TIMERDATA, &timerHandle);
```

So yeah I did allocate the timer on the heap, and would free it later once the app shuts down. The implementation of `httpStartTimeout` looks just like this: 

```
void httpCURLStartTimeout(CURLM* curlM, long timeout, void* userData) {
    if (timeout <= 0) timeout = 1;
    uv_timer_t* timer = (uv_timer_t*)userData;
    uv_timer_start(timer, httpCURLOnTimeout, timeout, 0);
}
```

It seems that it was able to get the timer from userData void pointer correctly, I just have no idea why there was a read access violation trying to access the associated event loop. I wasnt using the default event loop though, but switching to default loop did not solve the problem anyway.

**Reply by saghul** [reply]
`curl_multi_setopt(curlMultiHandle, CURLMOPT_TIMERDATA, &timerHandle);`

You probably want to do: `curl_multi_setopt(curlMultiHandle, CURLMOPT_TIMERDATA, timerHandle);`

**Reply by HallofFamer** [reply]
@saghul 

I see, thank you so much, it solved this problem I had with event loop in timer data. With this being said, I am now getting a different error: `Exception thrown: read access violation. h was 0xFFFFFFFFFFFFFFFF`, which points to the function `Curl_hash_pick`: 

```C
Curl_hash_pick(struct Curl_hash *h, void *key, size_t key_len)
{
  struct Curl_llist_element *le;
  struct Curl_llist *l;

  DEBUGASSERT(h);
  if(h->table) {
  ...
```

The offending code appears to be a call to `curl_multi_assign`, as I dont use global variable, the curl multi handle is assigned as a field on the CURLContext struct. Could this be a problem? Also in what circumstance a read access violation error may occur at `Curl_hash_pick`?

```
curl_multi_assign(context->curlM, socket, (void*)context);
```

**Reply by saghul** [reply]
I also use it the same way, with a context around it. Is the context alive for the whole duration of curl operations?

**Reply by HallofFamer** [reply]
> I also use it the same way, with a context around it. Is the context alive for the whole duration of curl operations?

Yeah the context is alive, though I added curl multi handle as a field in the context struct, not sure if this is a problem.

**Reply by saghul** [reply]
Shouldn't be... can you perhaps share the whole example in a gist?
